### PR TITLE
planner: demo of new cost interface(refactor)

### LINKED ISSUE #33945: planner: refactor the implementation of cost calculation
## Enhancement
The current implementation of cost calculation is **tightly coupled with physical optimization**. 
In physical optimization, the optimizer changes the plan structure frequently and updates the current cost correspondingly.
Here is an example:
```
func addPushedDownSelection(t *task) {
    tableReader := t.Plan.(*TableReader)
    pushedDownFilters := PushDownExprs(tableReader.filters)    // PHY-OPT
    selCost := t.Count() * CopCPUFactor                        // COST
    selection := NewSelection(pushedDownFilters)               // PHY-OPT
    selection.cost = selCost                                   // COST
    t.cost += selCost                                          // COST
    selection.Child = tableReader.Child                        // PHY-OPT
    tableReader.Child = selection.Child                        // PHY-OPT
}
```
`addPushedDownSelection` is used to push `selection` down, and comments `COST` and `PHY-OPT` are used to represent the corresponding line is for `cost calculation` or `plan change in physical optimization`, and then you can see they are coupled tightly.
The current implementation is hard to maintain and has already caused lots of issues: #32675 #32672 #32698 #27189 #30103 #32362.

To solve this problem thoroughly, we decided to refactor the implementation of cost calculation. 
A new interface `Plan.CalCost() float64` will be introduced and all code related to cost calculation will be removed from physical optimization, and then you can see the example above become like this：
```
func addPushedDownSelection(t *task) {
    tableReader := t.Plan.(*TableReader)
    pushedDownFilters := PushDownExprs(tableReader.filters)    // PHY-OPT
    selection := NewSelection(pushedDownFilters)               // PHY-OPT
    selection.Child = tableReader.Child                        // PHY-OPT
    tableReader.Child = selection.Child                        // PHY-OPT
}
```
And all code related to cost calculation will be moved into the new and standard interface `Plan.CalCost() float64`. 
Below is an example of `IndexReader`:
```
func (p *PhysicalIndexReader) CalPlanCost(taskType property.TaskType) float64 {
	p.planCost = p.indexPlan.CalPlanCost(property.CopSingleReadTaskType)   // COST: child's cost
	p.planCost += p.indexPlan.StatsCount() * getRowSize(p) * netFactor     // COST: net I/O cost
	p.planCost += getSeekCost(p)                                           // COST: net seek cost
	p.planCost /= float64(p.ctx.GetSessionVars().DistSQLScanConcurrency()) // COST: consider concurrency
	return p.planCost
}
```

After this refactoring, the cost model will become easier to maintain and calibrate.
And the cost of a plan becomes re-computable, which means you can invoke `Plan.CalCost()` multiple times in varied places, and I think some other modules like (`PlanCache`, `SPM`, `PlanRewriter`, `JoinReorder`, ...) can also get some benefits from it.

Here is [a detailed design doc](https://pingcap.feishu.cn/docs/doccnq4pL5HauSrwlJ0ExPb0trb)(Chinese) about this.
Here is [the demo](https://github.com/pingcap/tidb/pull/33563).

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: close #33945

Problem Summary:

### What is changed and how it works?
Please see https://github.com/pingcap/tidb/issues/33945 for more details.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: planner/core/find_best_task.go [hunk 1]
```diff
@@ -282,6 +282,11 @@ func (p *baseLogicalPlan) enumeratePhysicalPlans4Task(physicalPlans []PhysicalPl
 		if curTask.cost() < bestTask.cost() || (bestTask.invalid() && !curTask.invalid()) {
 			bestTask = curTask
 		}
+
+		//// TODO: introduce a new variable as a switch to control this
+		//if curTask.plan().CalPlanCost(property.RootTaskType) < bestTask.plan().CalPlanCost(property.RootTaskType) || (bestTask.invalid() && !curTask.invalid()) {
```
Comment (winoros) [inline planner/core/find_best_task.go:287]: It's possible that the task is not root type right?
Such as the case that double read case

Comment (qw4990) [reply planner/core/find_best_task.go]: You are right. Now I introduce a new function `getTaskType(task) type` to get their types.

File: planner/core/plan.go [hunk 2]
```diff
@@ -401,15 +412,34 @@ func (p *baseLogicalPlan) ExplainInfo() string {
 	return ""
 }
 
+type basePhysicalCost struct {
+	planCostInit bool
```
Comment (time-and-fate) [inline planner/core/plan.go:416]: Why do we need to cache them?

Comment (qw4990) [reply planner/core/plan.go:416]: For a plan `op1 -> op2 -> op3`, when generating this plan from bottom to top, `op1.CalCost()` can be invoked multiple times, so cache calculated costs to gain some performance benefits.

**Pull request closed without merging**